Library/Homebrew/utils.rb

```diff
@@ -89,6 +89,28 @@ def require?(path)
     raise unless e.message.include?(path)
   end
 
+  def output_stream(stream)
+    original_puts = method(:puts)
+    Kernel.define_method(:puts) do |*args|
+      stream.puts(*args)
+    end
+
+    original_print = method(:print)
+    Kernel.define_method(:print) do |*args|
```

Comment:
Instead of overriding methods how about doing an `instance_eval` on the passed block and provide `puts` and `print` methods?

Reply:
The problem with the `instance_eval` was when, for example in `help`, we call an entirely different function like `get_help_message`. Am I missing something here?

Reply:
@MikeMcQuaid a problem arises when we use that approach (unless I'm missing something). For example in `help.rb`, we have `puts path ? command_help(...)` which needs to be printed to `$stderr`. This method is `private` and I cannot call it if I am to use `instance_eval`.

Similarly, `generate_help_text` (in `cli/parser.rb`) would also be problematic in the `instance_eval` approach – hence the overriding.

Reply:
It could be defined from a variable outside the block. I do wonder if we're just better to have a `stderr_pretty_puts` and `stderr_pretty_puts ` or similar methods instead given these aren't used super often and to avoid the monkey patching.

Reply:
If the function is called outside the block, the `Tty` stream is still `$stdout`, so we face the original problem of the special characters being interpolated despite `$stderr` redirection. Seems like `stderr_pretty_puts` is the best way forward?

Reply:
To avoid the monkey patching, we can use *only* the second part I proposed, e.g. use this everywhere:

```ruby
Tty.with($stderr) do |stderr|
  stderr.puts Formatter.error("This is an error.")
end
```

Reply:
This should work 👍🏼.

Reply:
Works for me 👍🏻